apps/settings/js/panels/call/call_forwarding.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * Make sure all operations would be executed one by one.
+ *
+ * @module call/call_forwarding
+ */
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+  var DialogService = require('modules/dialog_service');
+  var CallUtils = require('panels/call/call_utils');
+  var CallConstant = require('panels/call/call_constant');
+
+  /**
+   * @class CallForwarding
+   * @returns {CallForwarding}
+   */
+  var CallForwarding = Module.create(function() {
+    // Stores current states (enabler or not) of the call forwaring reason
+    this._cfReasonStates = [0, 0, 0, 0];
+  });
+
+  CallForwarding.prototype._getConn = function(cardIndex) {
+    var conn = window.navigator.mozMobileConnections[cardIndex];
+    if (!conn) {
+      this.throw('We can\'t find the conn object on card ' + cardIndex);
+    } else {
+      return conn;
+    }
+  };
+
+  CallForwarding.prototype.setCallForwardingValues =
+    function(cardIndex, options) {
+      var promise = new Promise((resolve, reject) => {
+        var conn = this._getConn(cardIndex);
+        var key = options.key;
+        var number = options.number;
+        var enabled = !!options.enabled;
+        // While storing the settings into the database we avoid observing
+        // changes on those ones and enabling/disabling call forwarding.
+        //
+        // TODO we have to add this back later
+        if (!key) {
+          resolve();
+        } else if (this._cfReasonStates[
+            CallConstant.CALL_FORWARD_REASON_MAPPING[key]] === enabled) {
+          // Bails out in case the reason is already enabled/disabled.
+          resolve();
+        } else {
+          var mozMobileCFInfo = {};
+          mozMobileCFInfo.action = enabled ?
+            CallConstant.CALL_FORWARD_ACTION.REGISTRATION :
+            CallConstant.CALL_FORWARD_ACTION.DISABLE;
+          mozMobileCFInfo.reason =
+            CallConstant.CALL_FORWARD_REASON_MAPPING[key];
+          mozMobileCFInfo.serviceClass = conn.ICC_SERVICE_CLASS_VOICE;
+
+          if (!CallUtils.isPhoneNumberValid(number)) {
+            DialogService.alert('callForwardingInvalidNumberError', {
```

Comment:
I think we can simply return the error type and let the caller determine how to handle the error.